File: ruby/lesson 10/unique_paths_bottom_up.rb
# https://leetcode.com/problems/unique-paths/

def unique_paths(m, n)
    dp = Array.new(m + 1) { Array.new(n + 1) { 0 } }
    (1..m).each do |row|
        (1..n).each do |col|
            dp[row][col] = dp[row - 1][col] + dp[row][col - 1]
            dp[row][col] = 1 if row == 1 || col == 1
        end
    end
    dp[m][n]
end

File: javascript/lesson 10/uniquePathsBottomUp.js
// https://leetcode.com/problems/unique-paths/

function uniquePaths(m, n) {
    const dp = Array.from(Array(m + 1), () => new Array(n + 1).fill(0));
    for (let row = 1; row < dp.length; row++) {
        for (let col = 1; col < dp[row].length; col++) {
           dp[row][col] = (row == 1 && col == 1) ? 1 : dp[row - 1][col] + dp[row][col - 1];
        }
    }
    return dp[m][n];
}
